### PR TITLE
docs: add the error catalogue specification, plan, and task breakdown

### PR DESCRIPTION
**Pull request by ogenstad**
Specification, plan, and task breakdown for IFC-3034. No behaviour change: this adds `dev/specs/ifc-3034-error-catalogue/` covering how the SDK consumes Infrahub's GraphQL error catalogue, so that ordinary operations raise the specific error for the failure. `GraphQLError` remains the fallback for failures the catalogue does not cover and stays the common base class, so existing `except GraphQLError` code keeps working.

Ref: IFC-3034. Related: IFC-2279 (spike), INFP-468 (backend catalogue), GitHub #7498 (out of scope).

## Previously approved work

* feat: parse the server error envelope onto the base exception classes (approved in #1348)
* feat: unify and re-root the not-found classes under GraphQLError (approved in #1363)

## Decisions settled while drafting

- **A new `ApiError` base above both `AuthenticationError` and `GraphQLError`.** Authentication failures reach consumers from the REST path as well as GraphQL, so they cannot simply be re-rooted under `GraphQLError`. Verified that a 401/403 on a GraphQL call is already handled as an `httpx.HTTPStatusError` and raises `AuthenticationError` *before* the body is parsed for GraphQL errors — so `except GraphQLError` never caught auth failures, and no dual inheritance is needed to preserve compatibility.
- **`.code` is a catalogue string or `None`.** The `/api/...` envelope's `extensions.code` is an *integer* mirroring the HTTP status, a different thing with a different type; it is not surfaced through `.code`. The catalogue is GraphQL-only today.
- **Generated classes derive their parent from the declared HTTP status** — 401/403 under the authentication branch, everything else under `GraphQLError` — rather than a hand-maintained per-code mapping.
- **Infrahub generates the bindings into this repo as its `python_sdk` submodule**, matching how `protocols.py` and the generated schema models already arrive. No copy of the catalogue schema is vendored here, so there is one freshness invariant instead of two, policed by extending Infrahub's existing `validate-generated` check. No release-time gate is added on either side.
- **Query text is dropped from the message for catalogued errors only**; uncatalogued errors keep today's message verbatim.
- **`NodeNotFoundError`, `BranchNotFoundError` and `SchemaNotFoundError` are unified** with their catalogue counterparts and re-rooted under `GraphQLError`, accepting that `except GraphQLError` now also catches client-side lookup misses.

## Findings from the code survey worth a reviewer's eye

These are in the spec's Edge Cases section as specific hazards, not hypotheticals:

- **An ordered `isinstance` ladder gets shadowed.** `infrahub_sdk/ctl/utils.py:58-72` tests `GraphQLError` at line 67 *before* `(SchemaNotFoundError, NodeNotFoundError, ...)` at line 70. Re-rooting those classes makes the later branch unreachable, silently changing CLI output for exactly the errors this feature makes specific. FR-018 requires the correction.
- **A renderer with no server errors to render.** That same `GraphQLError` branch renders `exc.errors`, a list of server error dicts. A unified `NodeNotFoundError` raised purely client-side has no server response behind it, so the list is empty.
- **`identifier` carries two types.** The existing client-side `NodeNotFoundError` has `identifier` as a mapping of filters; the catalogue payload has it as a single string. FR-016 mandates the unification; the reconciliation mechanism is left to the plan.
- Eight more, including `NodeInvalidError` silently inheriting the re-rooting, a pre-existing call site passing a string where `GraphQLError` expects a list of error dicts, `UNDEFINED_ERROR` being a real code rather than the absence of one, and GraphQL data errors arriving as HTTP 200 while auth failures arrive as real 401/403 on a separate code path.

## Scope

Six prioritised user stories, 28 functional requirements. FR-025 to FR-027 land in the Infrahub repository (generation plus the extended drift check) and are tagged as such; everything else lands here.

`tasks.md` breaks the work into 83 tasks and maps them onto four pull requests: three here (the envelope with no bindings, the re-rooting, the typed per-code classes) and one in Infrahub (the generator). Phase numbering is a nominal sequence; only US3-before-US5 and US5-before-US1 are dependency-forced.

## Checks

`rumdl` clean across 131 files; Vale flags nothing in the new files. Requirements checklist at `dev/specs/ifc-3034-error-catalogue/checklists/requirements.md` passes 16/16 with no `[NEEDS CLARIFICATION]` markers remaining.

---
## Summary by cubic
Makes the SDK consume Infrahub's GraphQL error catalogue: authentication and GraphQL failures surface the server's error envelope, and the three lookup-miss codes now raise their own classes under `GraphQLError`. Also lands the IFC-3034 specification, plan, contracts, and 83-task breakdown in `dev/specs/ifc-3034-error-catalogue/`.

**What changed**
- `AuthenticationError` and `GraphQLError` share a new `ApiError` base carrying `code`, `http_status`, `extensions`, and `errors`, so callers can branch on the server's catalogue code instead of matching message text.
- Server-reported `NODE_NOT_FOUND`, `BRANCH_NOT_FOUND`, and `SCHEMA_NOT_FOUND` raise `NodeNotFoundError`, `BranchNotFoundError`, and `SchemaNotFoundError`; all three descend from `GraphQLError`, and `NodeInvalidError` inherits the re-rooting with `CODE` cleared.
- A catalogued failure's message now names the code and the server's message (`UNIQUENESS_VIOLATION: Node of kind TestPerson already has name 'John'`) instead of embedding the query text; uncatalogued failures keep the old message verbatim, and the full error list stays on `exc.errors`.
- `infrahub_sdk/exceptions.py` became a strictly layered package (`base`, `factory`, façade) with raise-time factories for every auth and GraphQL raise site; `infrahub_sdk.exceptions` stays the one supported import path, pinned by a snapshot test.
- A 401/403 whose body is not a readable error envelope raises `AuthenticationError` with the best reason available (server messages, then REST `detail`, then the plain status) instead of `JsonDecodeError`, `TypeError`, or the generic default message; the same parsing fixes an `AttributeError` on a valid non-object body, and a file 404 now always raises `NodeNotFoundError` whatever the body contains.
- `infrahubctl` prints the exception's message instead of "0 error(s)" for an unreadable GraphQL failure, and escapes bracketed text so rich no longer eats branch names.
- A client authenticating with an API token no longer retries after an expired-token 401, since the retry would replay the same rejected token.
- `NodeNotFoundError.identifier` is widened to `Mapping[str, list[str]] | str`, documenting a shape the SDK already raised.

**Design notes**
- The hierarchy is single-inheritance: the three 401/403 codes get no classes of their own, and the raised class comes from the response's first error code and the transport observed.
- `code_names_the_failure()` separates a described failure from a merely coded one: `UNDEFINED_ERROR` stays readable on `exc.code` but does not shape the message.
- The generated bindings, typed payload attributes, and the Infrahub-side generator land in later pull requests; `GraphQLError` remains the fallback for anything the catalogue does not cover.

<sup>Written for commit 2af53e7cc4fdcf5b5e83debfa2b2b7bf5f1ab8c9. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/opsmill/infrahub-sdk-python/pull/1266?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: dev/specs/ifc-3034-error-catalogue/tasks.md [hunk 1]
```diff
@@ -0,0 +1,549 @@
...
+
+## Path conventions
+
+Two checkouts are involved:
+
+- **SDK** (this repository): paths are relative to the repository root, e.g. `infrahub_sdk/exceptions/base.py`.
+- **Infrahub** (`opsmill/infrahub`, phase 7 only): paths are prefixed `[infrahub]`, e.g.
+  `[infrahub] tasks/backend.py`.
+
+## Phase order is dependency order, not priority order
```
[the remark this answers is not in the document]
Comment (ogenstad) [reply dev/specs/ifc-3034-error-catalogue/tasks.md]: Fixed in 9fe1d0d. The finding was correct, and the heading was the load-bearing sentence, so as written it would have serialized four independent phases.

The section now states that phase numbering is a nominal sequence and names the only two forced adjacencies: US3 before US5 (the CODE declarations are a prerequisite for generation succeeding at all, not just for its output) and US5 before US1. It says explicitly that US2, US3, US4 and US6 depend on Foundational alone and should not be serialized.

Two things beyond the suggestion. The parallel-opportunities bullet had the same defect in the other direction: it listed phases 3, 5 and 6 but omitted 4, and claimed the three share no files beyond the factory. US3 and US6 both add cases to test_exceptions.py, so both overlaps are now named. And the issue-split section listed 'issues 1 and 2 both touch base.py' among the dependencies, which it is not; it is now marked as a scheduling conflict rather than a constraint.
